### PR TITLE
add a random regular graph generator

### PR DESCRIPTION
**Pull request by marcliberatore**
This is a port of [NetworkX](https://github.com/networkx/networkx)'s [random regular graph generator](https://github.com/networkx/networkx/blob/master/networkx/generators/random_graphs.py) to Julia.

It's also some of the first non-trivial Julia code I've written, so you may want to eyeball it for non-Julian idioms.

### CODE REVIEW

File: src/randgraphs.jl [hunk 1]
```diff
@@ -81,3 +81,80 @@ function watts_strogatz(n::Integer, k::Integer, β::Real; is_directed=false)
     end
     return g
 end
+
+function random_regular_graph(n::Integer, k::Integer, seed=-1)
+    @assert(n * k % 2 == 0, "n * k must be even")
```
Comment (sbromberger) [inline src/randgraphs.jl]: why not use `iseven(n*k)`?

**Comment by jpfairbanks** [comment]
Thanks for contributing. I have two questions.

1. why did you define `suitable` and `try_creation` inside the definition of `random_regular_graphs`? Are they communicating through hidden state that can't be passed as arguments?

2. you have a `while !empty; try again` loop in there. Do we have a guarantee that it will terminate?

**Comment by sbromberger** [comment]
This is very cool, but in addition to @jpfairbanks - the performance could probably be improved:

```
julia> @time z = LightGraphs.random_regular_graph(1000000,4)
  37.812 seconds      (156 M allocations: 4850 MB, 38.89% gc time)
{1000000, 2000000} undirected graph
```

File: src/randgraphs.jl (hunk 1 above)
Comment (marcliberatore) [inline src/randgraphs.jl]: Ignorance. `iseven` seems like a better idea, now that I know about it.

**Comment by marcliberatore** [comment]
@jpfairbanks 
1. I was mirroring the nested structure from NetworkX — it's probably due to Python's "all-things-must-be-exported" behavior. I'll adjust the code to consist of only top-level functions.
2. I believe the algorithm is guaranteed to terminate. This is Steger and Wormald's "Algorithm A"; see <http://research.microsoft.com/en-us/um/redmond/groups/theory/jehkim/a-yspapers/grrg.pdf>.

@sbromberger You're right. That's awful performance. (My motivating use case was on graphs of about 10k nodes, so it wasn't a problem.)

I'll go after some of the simple stuff and see if I can improve things.

**Comment by sbromberger** [comment]
I have one that's not guaranteed to terminate (sort of like my other random graph generator `Graph(n,e)`) that's much faster, but if termination guarantee is required, we'll need to do something about it.

**Comment by sbromberger** [comment]
non-guaranteed version:

```
julia> @time y = LightGraphs.rrg2(1000000, 4)
   4.699 seconds      (6706 k allocations: 750 MB)
{1000000, 4000000} undirected graph
```

**Comment by sbromberger** [comment]
@marcliberatore I used your code and tightened it up a bit and now get

```
julia> @time z = LightGraphs.random_regular_graph(1000000,4)
   7.330 seconds      (14001 k allocations: 2096 MB, 12.63% gc time)
{1000000, 2000000} undirected graph
```

which I think is acceptable performance for a first cut. I'll paste the gist somewhere and you can have a look.

**Comment by marcliberatore** [comment]
It never fails to be humbling to see what someone who knows a language well can do with it.

I did bang on this last night for a while, adding type annotations and improving the use of memory in various spots, but the best I could do was about a 20% speedup. Clearly I have some things to learn. Thanks for the demonstration.

I can merge this PR, and/or I can squash the various commits together into one and force push to my fork if you like. Or you can just close this PR and adjust your repository directly. I'm not tied to being in the commit log or whatnot — I'm just trying to pass on code I found useful. Let me know what you prefer.

**Comment by sbromberger** [comment]
You should absolutely be in the commit log - I did nothing other than move the functions and do some type annotations (well, I changed things to use `Edge()` as well since that data structure is very nicely optimized)  :)

I created a PR on your local repo, so what I might suggest (and this is where git gets complicated for me) is that you merge my PR to your local repo, squash it, and I think that will automatically trigger a rebuilt PR here.

**Comment by sbromberger** [comment]
Also, if you have an optimization that I don't, include it :)

**Comment by sbromberger** [comment]
Erm - also, in `_suitable()`:

```
    for i in 1:length(ends)
        for j in i+1:length(ends)
            s1, s2 = ends[i:i+1]
            if s1 > s2
                s1, s2 = s2, s1
            end
            if ∉(Edge(s1, s2), edges)
                return true
            end
        end
    end
```
Why are we iterating `j` here when we don't use it, and won't we risk getting a BoundsError since we're using `i+1` when it iterates to the end of `ends`?

**Comment by sbromberger** [comment]
Waiting for a decision on https://github.com/marcliberatore/LightGraphs.jl/pull/2 by @marcliberatore before merging.

**Comment by marcliberatore** [comment]
1. If a user chooses a degree `k` that's greater than about 0.5 `n`, it's going to be faster to generate the graph of degree `n` - `k` and return its complement (when that approach is feasible). Should I add a case for this?
2. Should I update the user-facing documentation to note that this algorithm takes nk space and nk^2 time? And/or note the source of the algorithm?

**Comment by sbromberger** [comment]
Great questions.

1) Yes. We have `complement` in LightGraphs (see http://lightgraphsjl.readthedocs.org/en/latest/operators/).

2) I think we should indicate performance where possible in the docs.

**Comment by sbromberger** [comment]
PS: once this is done, @marcliberatore, do you think you could do a `random_regular_digraph()` generator?

**Comment by marcliberatore** [comment]
Please eyeball the last few commits. If you're happy with them, I can squash them pre-merge if you like.

Regarding `random_regular_digraph()`: This PR, or open a new one once I have something together?

**Comment by sbromberger** [comment]
Will do - thank you.

As far as `digraph` - after this. Let's not hold this up!

**Comment by sbromberger** [comment]
This is so cool. Thanks very much for your patience.

**Pull request merged**